src/lib/custom-resources.ts

```diff
@@ -0,0 +1,741 @@
+
+							pngFrames.push(imageData);
+							pngDelays.push(frameTimeInterpolated);
+						}
+					} else {
+						for (const frame of animation.frames) {
+							ctx.clearRect(0, 0, canvas.width, canvas.height);
+							ctx.drawImage(getFrame(image, frame.index), 0, 0);
+
+							pngDelays.push(frame.time);
+
+							const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height).data.buffer;
+
+							pngFrames.push(imageData);
+						}
+					}
+				}
+
+				if (pngFrames.length > 0) {
+					const apng = UPNG.encode(pngFrames, normalizedSize, normalizedSize, 0, pngDelays);
+
+					await fs.writeFile(textureFile, Buffer.from(apng));
+
+					try {
+						if (fs.existsSync(textureFile.replace('.png', '.gif'))) {
+							await execFile(apng2gif, [textureFile, '-o', textureFile.replace('.png', '.gif')]);
+						} else {
+							await execFile(apng2gif, [textureFile, '-o', textureFile]);
+						}
+					} catch (error) {
+						console.log(error);
+					}
+				}
+			}
+
+			pack.textures.push(texture as ItemTexture);
+		}
+	}
+}
+
+export function getPacks() {
+	return outputPacks;
+}
+
+export function getCompletePacks() {
+	return resourcePacks;
+}
+
+const textureMap = new Map();
+const allTextures = new Map();
+const textureIdDamageMap = new Map();
+const allTexturesIdDamage = new Map();
+const timeoutId = setTimeout(async () => {
+	if (!resourcesReady) {
+		await readyPromise;
+	}
+
+	for (const pack of resourcePacks) {
+		for (const texture of pack.textures) {
+			if (texture.skyblock_id !== undefined) {
+				const key = `${pack.config.id}:${texture.skyblock_id}`;
+				const data = textureMap.get(key) ?? [];
+
+				textureMap.set(key, [...data, texture]);
+				allTextures.set(texture.skyblock_id, true);
+				continue;
+			}
+
+			const itemId = texture.id;
+			const damage = texture.damage ?? 0;
+			if (itemId !== undefined && itemId !== 397) {
+				const key = `${pack.config.id}:${itemId}:${damage}`;
+				const data = textureIdDamageMap.get(key) ?? [];
+
+				textureIdDamageMap.set(key, [...data, texture]);
+
+				allTexturesIdDamage.set(`${itemId}:${damage}`, true);
+			}
+		}
+	}
+
+	for (const [key, value] of textureMap) {
+		textureMap.set(
+			key,
+			value.sort((a: ItemTexture, b: ItemTexture) => b.weight - a.weight)
+		);
+	}
+
+	for (const [key, value] of textureIdDamageMap) {
+		textureIdDamageMap.set(
+			key,
+			value.sort((a: ItemTexture, b: ItemTexture) => b.weight - a.weight)
+		);
+	}
+
+	clearTimeout(timeoutId);
+}, 100);
+
+/**
+ * Processes all textures that could potentially be connected to an item, then throws the one with biggest priority
+ * @param {object} item
+ * @param {object} options
+ * @param {boolean} [options.ignore_id]
+ * @param {string[]} [options.pack_ids]
+ * @param {boolean} [options.debug]
+ * @returns {object} Item's texture
+ */
+export function getTexture(item: SkyBlockItem, { ignore_id = false, pack_ids = [], debug = false, hotm = false } = {}) {
```

Comment:
Same as above, this function is also long and complex

Reply:
same as above